privval/signer_client_test.go

```diff
@@ -0,0 +1,245 @@
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerVoteKeepAlive(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			want := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+			have := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			// Check that even if the client does not request a
+			// signature for a long time. The service is will available
+			t.Log("TEST. Forced Wait")
+			time.Sleep(testTimeoutReadWrite * 2)
+			t.Log("TEST. Forced Wait - DONE")
+
+			require.NoError(t, tc.mockPV.SignVote(tc.chainID, want))
+			require.NoError(t, tc.signer.SignVote(tc.chainID, have))
+
+			assert.Equal(t, want.Signature, have.Signature)
+		}()
+	}
+}
+
+func TestSignerSignProposalErrors(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			// Replace service with a mock that always fails
+			tc.signerService.privVal = types.NewErroringMockPV()
+			tc.mockPV = types.NewErroringMockPV()
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			ts := time.Now()
+			proposal := &types.Proposal{Timestamp: ts}
+			err := tc.signer.SignProposal(tc.chainID, proposal)
+			require.Equal(t, err.(*RemoteSignerError).Description, types.ErroringMockPVErr.Error())
+
+			err = tc.mockPV.SignProposal(tc.chainID, proposal)
+			require.Error(t, err)
+
+			err = tc.signer.SignProposal(tc.chainID, proposal)
+			require.Error(t, err)
+		}()
+	}
+}
+
+func TestSignerSignVoteErrors(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			ts := time.Now()
+			vote := &types.Vote{Timestamp: ts, Type: types.PrecommitType}
+
+			// Replace signer service privval with one that always fails
+			tc.signerService.privVal = types.NewErroringMockPV()
+			tc.mockPV = types.NewErroringMockPV()
+
+			defer tc.signerService.OnStop()
+			defer tc.signer.Close()
+
+			err := tc.signer.SignVote(tc.chainID, vote)
+			require.Equal(t, err.(*RemoteSignerError).Description, types.ErroringMockPVErr.Error())
+
+			err = tc.mockPV.SignVote(tc.chainID, vote)
+			require.Error(t, err)
+
+			err = tc.signer.SignVote(tc.chainID, vote)
+			require.Error(t, err)
+		}()
+	}
+}
+
+type BrokenSignerDialerEndpoint struct {
+	*SignerDialerEndpoint
+}
+
+// nolint
+func (ss *BrokenSignerDialerEndpoint) writeMessage(msg RemoteSignerMsg) (err error) {
+	_, err = cdc.MarshalBinaryLengthPrefixedWriter(ss.conn, PubKeyResponse{})
+	return
+}
+
+func TestSignerUnexpectedResponse(t *testing.T) {
+	for _, tc := range getSignerTestCases(t) {
+		func() {
+			// TODO(jleni): This test is actually not working. Fails for a different reason
+
+			tc.signerService.privVal = types.NewErroringMockPV()
+			tc.mockPV = types.NewErroringMockPV()
+
+			// Replace signer service with a broken one
+			tc.signerService.OnStop()
+			tmp := BrokenSignerDialerEndpoint{tc.signerService}
+			tmp.OnStart()
```

Comment:
usually we don't call `OnStart` directly, but instead call Service'es `Start` func.